src/libraries/ZkAddress.sol

```diff
@@ -0,0 +1,157 @@
+// SPDX-License-Identifier: CC0-1.0
+
+pragma solidity 0.8.15;
+
+import "@base58-solidity/Base58.sol";
+
+/**
+ * @title ZkAddress
+ * Library for parsing zkBob addresses.
+ */
+library ZkAddress {
+    uint256 internal constant R = 21888242871839275222246405745257275088548364400416034343698204186575808495617;
+
+    error InvalidZkAddress();
+    error InvalidZkAddressLength();
+    error InvalidZkAddressChecksum();
+
+    struct ZkAddress {
+        bytes10 diversifier;
+        bytes32 pk;
+    }
+
+    /**
+     * @notice Parses zkBob address from the zkBob UI representation.
+     * Note that on-chain base58 decoding is quite gas intensive (610k gas),
+     * consider to use other gas efficient formats from the below.
+     * @param _rawZkAddress zk address base58 string representation in the zkBob UI format.
+     * @param _poolId id of the pool to verify checksum for.
+     */
+    function parseZkAddress(
+        string calldata _rawZkAddress,
+        uint24 _poolId
+    )
+        external
+        pure
+        returns (ZkAddress memory res)
+    {
+        bytes memory _rawZkAddressBytes = bytes(_rawZkAddress);
+        uint256 len = _len(_rawZkAddressBytes);
+
+        if (len > 63 || len < 47) {
+            revert InvalidZkAddressLength();
+        }
+
+        // _zkAddress == Base58.encode(abi.encodePacked(bytes10(diversifier_le), bytes32(pk_le), bytes4(checksum)))
+        bytes memory dec = Base58.decode(_rawZkAddressBytes);
+        if (_len(dec) != 46) {
+            revert InvalidZkAddressLength();
+        }
+        res = _parseZkAddressLE46(dec, _poolId);
+        if (uint256(res.pk) >= R) {
+            revert InvalidZkAddress();
+        }
+    }
+
+    /**
+     * @notice Parses zkBob address from the gas-efficient hex formats.
+     * Note difference in endianness among checksummed and non-checksummed formats.
+     * @param _rawZkAddress zk address hex representation in one of 3 formats.
+     * @param _poolId id of the pool to verify checksum for.
+     */
+    function parseZkAddress(bytes memory _rawZkAddress, uint24 _poolId) external pure returns (ZkAddress memory res) {
+        uint256 len = _len(_rawZkAddress);
+
+        if (len == 42) {
```

Comment:
Not always, but in this particular case I think it's pretty self explanatory, since the per-byte layout is described just one line below the constant, so the meaning of it is quite straightforward.